pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+	"sort"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/clock"
+)
+
+type backoffInfo struct {
+	key                      string
+	failuresAfterLastSuccess int32
+	lastFailureTime          *metav1.Time
+}
+
+type backoffInfoStore struct {
+	store cache.Store
+}
+
+func (s *backoffInfoStore) updateBackoffInfo(key string, failuresAfterLastSuccess int32, lastFailureTime *metav1.Time) {
+	b, ok, _ := s.store.GetByKey(key)
+
+	if !ok {
+		s.store.Add(&backoffInfo{
+			key:                      key,
+			failuresAfterLastSuccess: failuresAfterLastSuccess,
+			lastFailureTime:          lastFailureTime,
+		})
+	} else {
+		backoffInfo := b.(*backoffInfo)
+		backoffInfo.failuresAfterLastSuccess = failuresAfterLastSuccess
+		backoffInfo.lastFailureTime = lastFailureTime
+	}
+}
+
+func (s *backoffInfoStore) removeBackoffInfo(jobId string) {
+	b, exists, _ := s.store.GetByKey(jobId)
+	if exists {
+		s.store.Delete(b)
+	}
+}
+
+func newBackoffInfoStore() *backoffInfoStore {
+	return &backoffInfoStore{
+		store: cache.NewStore(backoffInfoKeyFunc),
+	}
+}
+
+var backoffInfoKeyFunc = func(obj interface{}) (string, error) {
+	if u, ok := obj.(*backoffInfo); ok {
+		return u.key, nil
+	}
+	return "", fmt.Errorf("could not find key for obj %#v", obj)
+}
+
+func getNewBackoffInfo(clock clock.WithTicker, backoffInfoStore *backoffInfoStore, key string, newSucceededPods []*v1.Pod, newFailedPods []*v1.Pod) backoffInfo {
+	currentTime := metav1.NewTime(clock.Now())
+	var oldBackoffInfo *backoffInfo
+	b, exists, _ := backoffInfoStore.store.GetByKey(key)
+	if exists {
+		oldBackoffInfo = b.(*backoffInfo)
+	}
+
+	if oldBackoffInfo == nil {
```

Comment:
nit: use `else`, then drop the `if oldBackoffInfo == nil {` condition.

Reply:
Done